### PR TITLE
ci: setup clang-format

### PR DESCRIPTION
**Pull request by SdtElectronics**
This PR is still WIP.
Status:
- [x] Add `.clang-format` file to configure `clang-format`
- [x]  Create a workflow for `clang-format` #99 
- [x]  Clean up current codebase using `clang-format` #100

### CODE REVIEW

**Review by cfriedt — CHANGES_REQUESTED** [review]
Looks like we may need a separate workflow for clang-tidy. I think that https://github.com/cpp-linter/cpp-linter-action might only be useful for wrapping around `clang-format` since it doesn't seem possible to include custom `CPPFLAGS` / `CXXFLAGS` for `clang-tidy` checks.

For `cpp-linter-action` with clang-format, I would suggest the following
* add a new `clang-format.yml` workflow
* if based on `cpp-linter-action`, specify
  * `tidy-checks: -*` (disable clang-tidy checking, that will go in a separate workflow)
  * `lines-changed-only`
  * `files-changed-only`
  
A separate `clang-tidy.yml` workflow should be relatively easy to do.

File: .clang-format [hunk 1]
```diff
@@ -0,0 +1,55 @@
+---
```
Comment (cfriedt) [inline .clang-format:1]: Looks like scancode requires a license statement at the top of this file.

**Comment by SdtElectronics** [comment]
> Looks like we may need a separate workflow for clang-tidy. I think that https://github.com/cpp-linter/cpp-linter-action might only be useful for wrapping around `clang-format` since it doesn't seem possible to include custom `CPPFLAGS` / `CXXFLAGS` for `clang-tidy` checks.
> 
> For `cpp-linter-action` with clang-format, I would suggest the following
> 
> * add a new `clang-format.yml` workflow
> * if based on `cpp-linter-action`, specify
>   
>   * `tidy-checks: -*` (disable clang-tidy checking, that will go in a separate workflow)
>   * `lines-changed-only`
>   * `files-changed-only`
> 
> A separate `clang-tidy.yml` workflow should be relatively easy to do.

`compile_flags.txt` serves the very purpose of providing compiler flags. It only supports setting flags for all files, though. Have gone through other `clang-tidy` actions in Github marketplace, I haven't find a single one supports passing flags as a parameter. We will have to customize a new workflow if we have more requirements on compiler flags.

A side note: `clang-tidy` supports designating flags for individual files, however, this is done with a compilation database, which is not meant to be handwritten, but generated by a build system. I tried using CMake to generate one for this project, but `clang-tidy` throws errors with that:
```
1 warning and 2 errors generated.
Error while processing /mnt/e/work_CS/embedded/zephyr-thrift-test/thrift-for-zephyr/lib/thrift/src/_stat.c.
error: unknown argument: '-fno-reorder-functions' [clang-diagnostic-error]
warning: optimization flag '-fno-defer-pop' is not supported [clang-diagnostic-ignored-optimization-argument]
/mnt/e/work_CS/embedded/zephyr-thrift-test/zephyr/lib/libc/newlib/include/stdint.h:15:10: error: 'newlib.h' file not found [clang-diagnostic-error]
#include <newlib.h>
         ^~~~~~~~~~
Found compiler error(s).
```

**Comment by cfriedt** [comment]
@SdtElectronics - similarly, you will need to rebase and edit your commit messages so that they each include a `Fixes #<issue>` item.

This has the effect of implicitly linking a commit against an issue. When the commit is merged, the issue is automatically closed.

To edit your commit messages, do this:
```shell
git rebase -i --autosquash 7f83cf985bf0e097fe93cb0736a006251d14f90a
# substitute "pick" with "reword" or simply "r"
# exit
# reword each commit to include `Fixes #<issue>` on a separate line.
```

**Comment by SdtElectronics** [comment]
CI failing due to `coding-style.sh` uses LGPL license. It is derived from the CERN Root project and the original license has to be preserved. On the other hand, this project doesn't depend on this script  to build or run, thus it should be fine to use it in a Apache 2.0 project.

File: .github/workflows/coding-style.yml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+      GH_PULL_REQUEST_BRANCH: ${{ github.head_ref }}
+      BASE_COMMIT: ${{ github.event.pull_request.base.sha }}
+    steps:
+    - uses: actions/checkout@v2
+    - name: Fetch base_ref HEAD
+      run: git fetch --depth=1 origin +refs/heads/${{github.base_ref}}:refs/remotes/origin/${{github.base_ref}}
+    - name: install clang-format
+      run: sudo apt-get install -y clang-format
+    - name: run coding-style script
+      run: scripts/coding-style.sh
```
Comment (cfriedt) [inline .github/workflows/coding-style.yml:23]: add a newline

Comment (cfriedt) [inline .github/workflows/coding-style.yml:23]: add a newline

File: scripts/coding-style.sh [hunk 3]
```diff
@@ -0,0 +1,30 @@
+# Copyright 2022 Henry Schreiner
+# SPDX-License-Identifier: LGPL-2.1-or-later
```
Comment (cfriedt) [inline scripts/coding-style.sh:2]: @SdtElectronics 
LGPL / GPL is incompatible with Apache-2.0 / Zephyr, so we can't use this script unfortunately.

I would suggest using this from upstream. It's Apache-2.0 licensed already 😁  Hopefully it works approximately the same

https://github.com/llvm/llvm-project/blob/c94a02e0e223bcce3b2c88741fae35d31a2d4f1d/clang/tools/clang-format/git-clang-format

**Review by cfriedt — CHANGES_REQUESTED** [review]
Update coding-style script to something permissively licensed.

File: samples/lib/thrift/hello_server/src/main.cpp [hunk 4]
```diff
@@ -27,44 +27,40 @@ using namespace ::apache::thrift::server;
...
 
-    int port = 4242;
-    ::std::shared_ptr<HelloHandler> handler(new HelloHandler());
-    ::std::shared_ptr<TProcessor> processor(new HelloProcessor(handler));
-    ::std::shared_ptr<TServerTransport> serverTransport(new TServerSocket(my_addr, port));
-    ::std::shared_ptr<TTransportFactory> transportFactory(new TBufferedTransportFactory());
-    ::std::shared_ptr<TProtocolFactory> protocolFactory(new TBinaryProtocolFactory());
-    TSimpleServer server(processor, serverTransport, transportFactory, protocolFactory);
+  int port = 4242;
+  ::std::shared_ptr<HelloHandler> handler(new HelloHandler());
```
Comment (cfriedt) [inline samples/lib/thrift/hello_server/src/main.cpp:47]: It looks as though it's much more common to use `std::` than `::std::` upstream. Let's stick with `std::` here.
```shell
cfriedt@cfriedt-mbp .upstream % grep -R "::std" * | wc -l
      65
cfriedt@cfriedt-mbp .upstream % grep -R "std" * | grep -v "::std" | wc -l
    8037
```

Looking much better already!

**Review by cfriedt — CHANGES_REQUESTED** [review]
replace of `::std::` with `std::`, for consistency

**Pull request merged**